### DISCUSSION TITLE
Detect if `this` is modified within method

### DISCUSSION THREAD

**Discussion by tqwewe** [swc-project/swc#3205] [Q&A]
I am building an event sourcing system, and need all methods to have a readonly instance of `this`.. but one method called `handle` can mutate `this`.

I am hoping to do this by parsing the AST and detecting if `this` is being modified in the method, but have no idea if I could reliably do something like this in swc.

Any suggestions would be greatly appreciated 👍

**Comment by kdy1** [comment]
Are you going to use rust?

**Reply by tqwewe** [reply]
Yep, my backend is in Rust. I am using TypeScript as a kind of scripting language for the end user. I just want to ensure they will not modify `this` in their methods they define. (except the method called `handle`)

**Reply by kdy1** [reply]
Then I think it will be simple.

Rustdoc of `swc` contains lots of useful links for rust-side users. https://docs.rs/swc/latest/swc/

**Reply by tqwewe** [reply]
Would you suggest I use the AST to verify this? I'm just thinking it will be quite difficult to catch situations with `Object.assign(this, ...)` and other ways to modify `this`.

**Reply by tqwewe** [reply]
Another way, probably more simple would be to clone `this`, and `Object.freeze` it before calling the users method

**Reply by kdy1** [reply]
Yeah, it will be quite hard if you want to perfectly block users from modifying this.
Users can pass `this` using arguments and if then there will be no usage of `this`
